codelists/actions.py

```diff
@@ -491,30 +492,116 @@ def export_to_builder(*, version, author):
 
-    cache_hierarchy(version=draft)
+def _check_new_concepts(draft):
+    """
+    Check for new matching concepts and ensure the codeset is up to date so the
+    builder frontend can display it
+
+    The recreated search will add any new concepts that are returned by a search.
+    The builder frontend can't deal with any new concepts (with unresolved status)
+    if they have a parent that is included/excluded.
+
+    Update all code statuses with the known statuses that are explicitly included
+    or excluded.  This will assign statuses to unresolved codes that have included/excluded
+    parents.  It will also update any previously implicitly included/excluded status that
+    are no longer applicable.
+
+    e.g. On the initial draft, B inherited from A;
+     - A was explicitly included, so B was implicitly included
+    After a coding system update, now A and B have swapped, so A inherits from B
+     - A is still explicitly included
+     - B is now unresolved as it has no parents that are explicitly included or
+       excluded (the fronted can deal with that)
+    """
+    # Make a copy of the pre-update code status; this will contain all codes returned
+    # by the recreated searches.  New codes will have status "?" at this stage, and
+    # any old codes will still be on the draft
+    initial_statuses = {**draft.codeset.code_to_status}
+
+    # if there are any code objs on the draft that are no longer in the coding system,
+    # delete them
+    draft_codes = set(draft.code_objs.values_list("code", flat=True))
+    codes_in_coding_system = set(draft.coding_system.code_to_term(draft_codes))
+    old_codes = draft_codes - codes_in_coding_system
+    if old_codes:
+        CodeObj.objects.filter(version=draft, code__in=old_codes).delete()
+        logger.info(
+            "Old codes no longer in coding system removed from draft",
+            draft=draft,
+            coding_system=draft.coding_system,
+            removed_codes=old_codes,
+        )
 
-    return draft
+    # Update the code statuses again for the explicitly included/excluded codes, and
+    # reset any implicit statuses
+    explicitly_included_or_excluded = {
+        (code, "+") for code in draft.codeset.codes("+")
+    } | {(code, "-") for code in draft.codeset.codes("-")}
+    builder_actions.update_code_statuses(
+        draft=draft, updates=explicitly_included_or_excluded, reset=True
+    )
+
+    # return a dict of new, removed and changed codes
+    all_changes = set(draft.codeset.code_to_status.items()) - set(
+        initial_statuses.items()
+    )
+    new_code_statuses = {
+        (code, status)
+        for code, status in draft.codeset.code_to_status.items()
+        if initial_statuses[code] == "?"
+    }
+    changed_code_statuses = all_changes - new_code_statuses
+    return {
+        "added": new_code_statuses,
+        "changed": changed_code_statuses,
+        "removed": old_codes,
+    }
```

Comment:
I would prefer it if this behaviour was captured in a couple of new methods on a codeset:

```
class Codeset:
    ...

    def update_hierarchy(self, new_hierarchy):
        ...
        return new_codeset

    def diff(self, other_codeset):
        ...
        return {
            "added": new_code_statuses,
            "changed": changed_code_statuses,
            "removed": old_codes,
        }
```

I feel like this would be easier to understand/test/maintain, which would offset a slightly less efficient implementation.

(The new `test_update` tests are very good, but I think they'd be easier to understand if they were tests against new `Codeset` methods.)

Reply:
I agree with putting this in methods on a codeset, but I'm unsure what you intended `update_hierarchy` on a `Codeset` to do.  At this point, the hierarchy has already been updated (with `cache_hierarchy`, after the searches are rerun) - I don't think that belongs in `update_hierarchy`, because it needs the `CodelistVersion` and its searches.   Would `update_hierarchy` just call the existing `Codeset.update` method instead of having the `reset` flag I added to it?  Or did you intend this to do the part where it deleted CodeObjs that no longer exist?
Sorry if this should be obvious

Reply:
Actually, I don't think the deleting CodeObjs bit works as a Codeset method either, because it's using the draft to filter the CodeObjs as well as the draft's current codeset.  And then it uses `builder_actions.update_code_statuses` to create the new codeset on the draft, which calls the codeset's `update` method and then updates the relevant statuses of the draft's CodeObjs.  Currently a Codeset doesn't have any knowledge itself of a CodelivstVersion.